### PR TITLE
samples: net/vlan: add IPv6 prefix config to each vlan-iface

### LINKED ISSUE #77401: tcp/ip: ipv6 prefix is not configured by default
**Describe the bug**

take sample/net/vlan as example, the ipv6 is null, see below `net status` log

```
IPv6 prefixes (max 2):
        <none>
```

**Steps to reproduce**
```
west build -b mimxrt1170_evk/mimxrt1176/cm7
west flash
```

**Expected behavior**

the ipv6 prefix shall has a default value, however it is empty. If an IPv6 Global-unicast-address is set, the IPv6-prefix should also be set accordingly.

**Impact**

ipv6 routing function is implement wrongly.

**Logs and console output**

```
Interface ? (0x80000b50) (Virtual) [2]
==================================
Virtual name : VLAN to 1
Attached  : 1 (Ethernet / 0x80000a40)
Link addr : 02:04:9F:D1:36:55
MTU       : 1500
Flags     : NO_AUTO_START,IPv4,IPv6
Device    : VLAN_0 (0x60025758)
VLAN tag  : 100 (0x064)
IPv6 unicast addresses (max 5):
        2001:db8:100::1 manual preferred infinite
        fe80::4:9fff:fed1:3655 autoconf preferred infinite
IPv6 multicast addresses (max 5):
        ff02::1
        ff02::1:ff00:1
        ff02::1:ffd1:3655
**IPv6 prefixes (max 2):
        <none>**
IPv6 hop limit           : 64
IPv6 base reachable time : 30000
IPv6 reachable time      : 32435
IPv6 retransmit timer    : 0
IPv4 unicast addresses (max 1):
        192.168.102.111/255.255.255.0 manual preferred infinite
IPv4 multicast addresses (max 2):

```

**Environment (please complete the following information):**

 - OS: (e.g. Linux, MacOS, Windows)
 - Toolchain (e.g Zephyr SDK, ...)
 - Commit SHA or Version used: v3.7.0-1337-gf93e19276e18

**Additional context**

a simple trial is add below for samples/net/vlan application

adding below lines, after the 

```
                if (!net_if_ipv6_prefix_add(vlan, &addr6, 64, NET_IPV6_ND_INFINITE_LIFETIME)) {
                        LOG_ERR("Cannot add %s with prefix_len %d to interface %p",
                                        ipv6_addr, 64, vlan);
                        return -EINVAL;
                }

```
after the 
```
                ifaddr = net_if_ipv6_addr_add(vlan, &addr6,
                                              NET_ADDR_MANUAL, 0);
                if (!ifaddr) {
                        LOG_ERR("Cannot add %s to interface %p",
                                ipv6_addr, vlan);
                        return -EINVAL;
                }

```

then the net status will show below

```
Virtual name : VLAN to 1
Attached  : 1 (Ethernet / 0x80000a40)
Link addr : 02:04:9F:20:88:38
MTU       : 1500
Flags     : NO_AUTO_START,IPv4,IPv6
Device    : VLAN_0 (0x60025744)
VLAN tag  : 100 (0x064)
IPv6 unicast addresses (max 5):
        2001:db8:100::1 manual preferred infinite
        fe80::4:9fff:fe20:8838 autoconf preferred infinite
IPv6 multicast addresses (max 5):
        ff02::1
        ff02::1:ff00:1
        ff02::1:ff20:8838
IPv6 prefixes (max 2):
        2001:db8:100::1/64 infinite
IPv6 hop limit           : 64
IPv6 base reachable time : 30000
IPv6 reachable time      : 44682
IPv6 retransmit timer    : 0
IPv4 unicast addresses (max 1):
        192.168.102.111/255.255.255.0 manual preferred infinite
IPv4 multicast addresses (max 2):
        224.0.0.1
IPv4 gateway : 0.0.0.0

```

### PR DESCRIPTION
**Pull request by shrek-wang**
Without the IPv6-prefix set, the net_if_ipv6_addr_onlink() check will always fail. The packets with DIP in the local link will go to 'try_send' code branch, which is not expected.

Fixes: #77401

### CODE REVIEW

File: include/zephyr/net/net_ip.h [hunk 1]
```diff
@@ -745,6 +745,39 @@ static inline bool net_ipv6_is_prefix(const uint8_t *addr1,
...
+
+/**
+ * @brief Get the IPv6 network address via the unicast address and the prefix mask.
+ *
+ * @param inaddr Unicast IPv6 address.
+ * @param outaddr Prefix masked IPv6 address (network address).
+ * @param length Prefix length (max length is 128).
+ */
+static inline void net_ipv6_addr_prefix_mask(const uint8_t *inaddr,
+		uint8_t *outaddr, uint8_t prefix_len)
```
Comment (jukkar) [inline include/zephyr/net/net_ip.h:757]: Could you align the parameters like this
```
static inline void net_ipv6_addr_prefix_mask(const uint8_t *inaddr,
					     uint8_t *outaddr,
					     uint8_t prefix_len)
```

File: include/zephyr/net/net_ip.h [hunk 2]
```diff
@@ -745,6 +745,39 @@ static inline bool net_ipv6_is_prefix(const uint8_t *addr1,
...
 }
 
+
+/**
+ * @brief Get the IPv6 network address via the unicast address and the prefix mask.
+ *
+ * @param inaddr Unicast IPv6 address.
+ * @param outaddr Prefix masked IPv6 address (network address).
+ * @param length Prefix length (max length is 128).
+ */
```
Comment (jukkar) [inline include/zephyr/net/net_ip.h:755]: Please place this function in a separate commit (we do not want to mix sample changes to other network changes in the same commit).

File: samples/net/vlan/src/main.c [hunk 3]
```diff
@@ -63,6 +63,19 @@ static int setup_iface(struct net_if *iface, struct net_if *vlan,
...
+		net_ipv6_addr_prefix_mask((uint8_t *)&addr6, (uint8_t *)&netaddr6,
+					CONFIG_NET_SAMPLE_IFACE_MY_IPV6_PREFIXLEN);
+
+		if (!net_if_ipv6_prefix_add(vlan, &netaddr6,
+					CONFIG_NET_SAMPLE_IFACE_MY_IPV6_PREFIXLEN,
+					(uint32_t)0xffffffff)) {
+			LOG_ERR("Cannot add %s with prefix_len %d to interface %p",
+					ipv6_addr,
+					CONFIG_NET_SAMPLE_IFACE_MY_IPV6_PREFIXLEN,
+					vlan);
```
Comment (jukkar) [inline samples/net/vlan/src/main.c:76]: Could you align the parameters in these functions

**Comment by jukkar** [comment]
No need to keep this as a draft, the changes look fine and can be reviewed already.

File: include/zephyr/net/net_ip.h [hunk 4]
```diff
@@ -745,6 +745,38 @@ static inline bool net_ipv6_is_prefix(const uint8_t *addr1,
 	return (addr1[bytes] & mask) == (addr2[bytes] & mask);
```
Comment (jukkar) [inline include/zephyr/net/net_ip.h:745]: Please change the commit subject of the 1st one to `net: Add ipv6 prefix mask function`

File: samples/net/vlan/Kconfig [hunk 5]
```diff
@@ -53,4 +53,12 @@ config NET_SAMPLE_IFACE3_VLAN_TAG
 	  Set VLAN (virtual LAN) tag (id) that is used in the sample
```
Comment (jukkar) [inline samples/net/vlan/Kconfig:53]: and the commit subject of the second one to `samples: net: vlan: Add IPv6 prefix config to each vlan-iface`

**Review by jukkar — COMMENTED** [review]
The signed-off-by has error, the compliance checker says
```
1: UC2 Signed-off-by: must have a full name
```
for both commits.

**Comment by shrek-wang** [comment]
All done. Thanks, @jukkar :)

**Review by jukkar — DISMISSED** [review]

**Review by rlubos — DISMISSED** [review]

**Comment by jukkar** [comment]
Some fixed need for documentation (seen in CI)
```
WARNING: /home/runner/_work/zephyr/zephyr/include/zephyr/net/net_ip.h:752: argument 'length' of command @param is not found in the argument list of net_ipv6_addr_prefix_mask(const uint8_t *inaddr, uint8_t *outaddr, uint8_t prefix_len)
WARNING: /home/runner/_work/zephyr/zephyr/include/zephyr/net/net_ip.h:752: The following parameter of net_ipv6_addr_prefix_mask(const uint8_t *inaddr, uint8_t *outaddr, uint8_t prefix_len) is not documented:
WARNING: /home/runner/_work/zephyr/zephyr/include/zephyr/net/net_ip.h:752: argument 'length' of command @param is not found in the argument list of net_ipv6_addr_prefix_mask(const uint8_t *inaddr, uint8_t *outaddr, uint8_t prefix_len)
WARNING: /home/runner/_work/zephyr/zephyr/include/zephyr/net/net_ip.h:752: The following parameter of net_ipv6_addr_prefix_mask(const uint8_t *inaddr, uint8_t *outaddr, uint8_t prefix_len) is not documented:
```

**Review by rlubos — APPROVED** [review]

**Comment by shrek-wang** [comment]
> Some fixes need for documentation (seen in CI)
> 
> ```
> WARNING: include/zephyr/net/net_ip.h:752: argument 'length' of command @param is not found in the argument list of net_ipv6_addr_prefix_mask(const uint8_t *inaddr, uint8_t *outaddr, uint8_t prefix_len)
> WARNING: include/zephyr/net/net_ip.h:752: The following parameter of net_ipv6_addr_prefix_mask(const uint8_t *inaddr, uint8_t *outaddr, uint8_t prefix_len) is not documented:
> WARNING: include/zephyr/net/net_ip.h:752: argument 'length' of command @param is not found in the argument list of net_ipv6_addr_prefix_mask(const uint8_t *inaddr, uint8_t *outaddr, uint8_t prefix_len)
> WARNING: include/zephyr/net/net_ip.h:752: The following parameter of net_ipv6_addr_prefix_mask(const uint8_t *inaddr, uint8_t *outaddr, uint8_t prefix_len) is not documented:
> ```

**Comment by shrek-wang** [comment]
@jukkar , I've fixed the argument to the right one.

**Review by jukkar — APPROVED** [review]

**Pull request merged**